### ISSUE TITLE
SMILESLanguage transforms can't be set in datasets

### ISSUE THREAD

**Issue by jannisborn** [PaccMann/paccmann_datasets#103]
If a SMILESLanguage or SMILESTokenizer object is passed to DrugAffinityDataset or any downstream class like SMILESTokenizerDataset, it is impossible to control the transforms of the tokenizer object. This is only possible if a new object is created.
See [SMILESTokenizerDataset](https://github.com/PaccMann/paccmann_datasets/blob/4f37cf3234598c22a8966c9402a1f19c69d569f5/pytoda/datasets/smiles_dataset.py#L144):

```py
        if smiles_language is not None:
            self.smiles_language = smiles_language
        else:
            language_kwargs = {}  # SMILES default
            if selfies:
                language_kwargs = dict(
                    name='selfies-language', smiles_tokenizer=split_selfies
                )
            self.smiles_language = SMILESTokenizer(
                **language_kwargs,
                canonical=canonical,
                augment=augment,
                kekulize=kekulize,
                all_bonds_explicit=all_bonds_explicit,
```

I suggest to do as follows:
- add a line above in the first `if` where `self.smiles_language.set_smiles_transforms()` is called to set the transforms with what has been passed to the SMILESTokenizerDataset constructor
- More importantly: Find a way to save the transforms as part of the SMILES Tokenizer object. These transformations can heavily indicate performance. At this point, it is **very** easy to mess things up. Effectively whenever any transformation was different from the default during model training AND a smiles language is passed at test time when the model is evaluated e.g. on a different dataset, the language will behave differently. That's a bug.
- I believe we should wrap all data needed to construct a smiles language/tokenizer into a _single_ `.json`. Now we keep dragging 3 files (vocab, `tokenizer_config` and `token_count`) from which effectively only the first is needed for restoring. If we unify these into a nested json that also has a logic to restore the transforms we heavily simplify our lives.

**Comment by jannisborn** [comment]
Plus, I think we should raise a warning whenever a user passes a  SMILESLanguage, restores the transformations from the pickle and then overwrites them to something different with the arguments to the dataset class.

**Comment by jannisborn** [comment]
Plus, it probably makes sense to drag 2 sets of transforms, one for train and one for testing. Those could be empty in SMILESLanguage but loaded in SMILESTokenizer and should be separately in the language `json`. Usually during training of predictive models, augmentation will be on, but not so during testing. But during testing we usually need to canonicalize since the augmentation does this implicitly (especially if the dataset is not consistent in terms of kekulization. This can be controlled via `kekule_smiles`, but at the end of the day most users wont notice this and it can introduce huge effects).

We should then be able to switch between the two sets of transforms by `smiles_language.train()` and `.eval()` to stick with pytorch conventions.

**Comment by C-nit** [comment]
There is a lot in here. I'll try to address the different concerns.
- **passing a smiles_language to the dataset vs constructing it internally**.
- **loading a vocab (any SmilesLanguage) vs restoring a pretrained instance of a specific class**. 
- **multiple instances e.g. for train/test.**. 


1. **passing a smiles_language to the dataset vs constructing it internally**.
I agree this is bad. But only because it is not clear at all how the class behaves depending on the input parameters. That most arguments are ignored when passing a smiles_language instance is not obvious, however well documented.  
The **solution** to this is to force creation of the smiles_instance outside of the dataset entirely. This is the common use case anyway! Anyway constructing a new one inside a specific dataset is a bad idea, as you would want to add also validation and test .smi to it (a complete vocab) and use the same vocab/language in different datasets.  
This will make the docs of one of our most used datasets much more readable and clearer that any implementation of SmilesLanguage can be used, but has to be managed outside of the dataset (I think this addresses the bug you mention.)
We will also not require the warning you rightfully suggest for the current class.

2. **loading a vocab (any SmilesLanguage) vs restoring a pretrained instance of a specific class**.
`class.from_pretrained` (here you need to know the class) allows restoring of a specific instance, using original init arguments also restoring token counts and max_squence_length (which can be relevant for padding). Note that you can overwrite any init arguments: `smiles_language = class.from_pretrained(pretrained_path, canonical=True)`.
However, noone forces you to restore full instances anymore since we have the vocab stored separate. Create any instance, then load only the vocab. So it is very important to have the vocab separate.

3. **multiple instances e.g. for train/test.**
First, I don't think we should use the same dataset for different settings (.tain()/.eval() on the dataset/language). There is a datasets for training with fixed behaviour. One can of course change it, but this is not the design.  

The scenario is that we create a smiles_language to train a model. We need to store things in a way we can do inference on that model later.
- The most important thing is that the inference dataset uses the same vocab, else the inference is nonsensical. This is covered both by `instance.load_vocabulary` and `class.from_pretrained`, so I consider this covered.  
- Just as important is to carry/drag transformation arguments that affect the shape. For example changing `padding_length` would fail on inference. This is solved by `from_pretrained` out of the box, and can be done manually with `load_vocabulary`. In both cases we can also use matching parameters that were stored/dragged in some other fashion, e.g. model params. But that the parameters are correct is responsibility of the user in that case. That is the benefit of `from_pretrained`.
- Now some parameters might be allowed to change or have to change in the inference case.  
The point here would be to restore a smiles_language known to work from training, then change parameters you know to change.
  Personally I would still have only one version of pretrained or vocab stored from training. I would fix these specific inference params in some fashion -could be hardcoded in the script even- and use **only** them for overwriting the args of the smiles_language from training (as shown in 2. passing to `from_pretrained`).

**Comment by jannisborn** [comment]
Thanks @C-nit. Came across this in the cleanup. You convinced me here, especially regarding 2) and 3). It doesnt seem to be a big blocker given the silence in the last year.

Regarding 1): What I'm doing now is simply an improved logging that, if a SMILES language was passed, informs the user  that all the arguments will be ignored. Then I loop over all affected arguments (like `canonical`) and check whether there's a mismatch between the provided `canonical` and the `smiles_language.canonical`. If there's a mismatch, I throw a logging error that the smiles_language takes preference and instruct how to fix the issue. I agree that `SMILESTokenizerDataset` does not need to be able to change an existing language object according to the user needs. This should happen outside the class.
